Support Traceflow of live traffic

jianjuns opened the pull request:
Add support of tracing live traffic. Rather than injecting a Traceflow
packet, a live traffic Traceflow will trace the real traffic between
Pods - the first packet of the first connection that matches the
Traceflow spec will be traced.
antctl traceflow command is extended to support live traffic Traceflow.
This commit also makes a few others changes to Traceflow: change
Destination (live traffic Traceflow does not require Traceflow
Destination to be provided), Packet, and Packet IPHeader in Traceflow
Spec to optional fields; add a Timeout parameter to Traceflow Spec and
antctl traceflow command to specify the timeout time of a Traceflow;
delete OVS flows added for the Traceflow after agent receives the first
captured packet of the Traceflow; support all IP protocols.

jianjuns commented:
I still need to add e2e test, and fix the OVS packet_out error for regular Traceflow (at least in my env).

@antoninbas : I plan to add reporting the captured packet headers after this PR.

gran-vmv commented on pkg/ovs/openflow/ofctrl_builder.go:
```diff
@@ -488,6 +499,19 @@ func (b *ofFlowBuilder) MatchDstPort(port uint16, portMask *uint16) FlowBuilder
 
+// MatchDstPort adds match condition for matching source port in transport layer. OVS will match the port exactly
+// if portMask is nil.
+func (b *ofFlowBuilder) MatchSrcPort(port uint16, portMask *uint16) FlowBuilder {
```
MatchTcpSrcPort?

jianjuns replied:
Thanks!

gran-vmv commented on pkg/antctl/raw/traceflow/command.go:
```diff
@@ -127,19 +144,19 @@ func runE(cmd *cobra.Command, _ []string) error {
 	}
 	defer func() {
-		if option.waiting {
+		if !option.nowait {
```
Could you explain about this change?

jianjuns replied:
I hoped to keep the boolean options consistent and the default value to false. But if there are other cases to use waiting, I can revert the change.

gran-vmv replied:
OK. Just want to know the purpose of this change.
@lzhecheng Any comments here?

lzhecheng replied:
No strong opinion here.

gran-vmv commented on pkg/antctl/raw/traceflow/command.go:
```diff
@@ -249,13 +270,16 @@ func parseFlow() (*v1alpha1.Packet, error) {
 	}
 
 	pkt := new(v1alpha1.Packet)
+	// TODO: support IPv6.
```
Lan Luo added IPv6 support in Antrea 1.0: https://github.com/vmware-tanzu/antrea/pull/1995

jianjuns replied:
Great to know. Removed to the TODO.

gran-vmv commented on pkg/antctl/raw/traceflow/command.go:
```diff
@@ -179,34 +196,37 @@ func newTraceflow(client kubernetes.Interface) (*v1alpha1.Traceflow, error) {
+			} else {
+				return nil, fmt.Errorf("destination should be in the format of Namespace/Pod, Pod, Namespace/Service or Service")
+			}
+			if isPod, err = dstIsPod(client, dst.Namespace, dest); err != nil {
```
Do you think we can use another parameter for destination service?

gran-vmv replied:
If pod and service share same parameter, the antctl cannot support trace a service which have same name with pod. I'm OK if we can accept this restriction.

gran-vmv replied:
Got it.

jianjuns commented on pkg/agent/controller/traceflow/traceflow_controller.go:
```diff
@@ -312,144 +364,176 @@ func (c *Controller) validateTraceflow(tf *opsv1alpha1.Traceflow) error {
+			}
+
+			if !liveTraffic {
+				// Why need to set TCP flags to 2?
```
@gran-vmv : why we need to set TCP flags to 2 for only Service? I hope to add a comment here.

gran-vmv replied:
For service, it will be DNATed, thus SYN flag is required. For other traffic, it will bypass traceflowConntrachFlow.
I think we need to set the default TCP flag to 2.

antoninbas commented on pkg/apis/ops/v1alpha1/types.go:
```diff
@@ -103,9 +103,15 @@ type Traceflow struct {
-	Destination Destination `json:"destination,omitempty"`
-	Packet      Packet      `json:"packet,omitempty"`
+	Source      Source       `json:"source,omitempty"`
+	Destination *Destination `json:"destination,omitempty"`
+	Packet      *Packet      `json:"packet,omitempty"`
```
I do not believe this is considered a backwards-compatible change unfortunately, so the API version needs to change. A Go client for example would need to be updated.

jianjuns replied:
I do feel these changes are compatible. For example, I can test with old antctl + old controller + new agent.

antoninbas replied:
The serialized data is the same indeed. Let me see if I find something in the K8s API documentation about changing from non-pointer to pointer type.

jianjuns replied:
Sure. Let me know. If it is not compatible, I might choose to revert the changes.

antoninbas replied:
All I found with a quick search was [this](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#on-compatibility):

> An API change is considered compatible if it [...] does not change existing semantics, including [...] which fields are required and which are not

And later on in the same document, when discussing upgrading alpha APIs:

> developers should still increment the API version when object schema or semantics change in an incompatible way

I think typically making a field optional would be considered incompatible because a client may list all resources and fail to process them (e.g. segfault) if a field that was previously required is now optional (i.e. became a pointer). Because of the nature of the Traceflow API, its current (known) consumers, and the usage these consumers make of the API, I would agree that it is an unlikely situation. If I am writing a Go client which uses an API like Traceflow, it is surprising that the package remains exactly the same (`pkg/apis/ops/v1alpha1`), yet my code has to change if I want to be able to keep building it with the latest imported code. If we do want to apply the guidelines for API versioning, it seems to me we should still increase the version to v1alpha2. What do you think @tnqn?

jianjuns replied:
I feel changing from required to not is better...

I do not know if code level compatibility is also covered by API version. Like to learn.

antoninbas replied:
> I feel changing from required to not is better...

For a POST yes, but for a GET?

> I do not know if code level compatibility is also covered by API version. Like to learn.

I feel like they go hand-in-hand here, if you consider API reads in addition to writes

jianjuns replied:
Ok. Let us see what @tnqn thinks. Given we are making APIGroup changes and there are only several days left for the release, I would just revert the changes if we need to add a new API version for them.

tnqn replied:
> I do feel these changes are compatible. For example, I can test with old antctl + old controller + new agent.

The scenario we need to support is new controller + old agent? I'm thinking a case that new antctl create a Traceflow that doesn't set Destination and Packet, would agent crash or the optional fields will be filled up with zero value?
As @antoninbas's link pointed out, changing existing fields' semantics in this way is considered incompatible and is generally disallowed. I don't know what it would break except the code consuming it. But if it's just good to make it optional, not must, I feel it's better to defer the change to next API version to be riskless.

jianjuns replied:
I think old agent will not crash, as its go client will populate an empty struct.

Ok, as we believe generally changing to optional is an incompatible change, let me drop it in this release.

jianjuns commented on pkg/agent/controller/traceflow/packetin.go:
```diff
@@ -89,17 +90,28 @@ func (c *Controller) parsePacketIn(pktIn *ofctrl.PacketIn) (*opsv1alpha1.Tracefl
+	c.runningTraceflowsMutex.RUnlock()
+
+	if firstPacket {
+		// Uinstall the OVS flows after receiving the first packet.
```
Fixed. Thanks!

jianjuns commented on pkg/agent/controller/traceflow/packetin.go:
```diff
@@ -89,17 +90,28 @@ func (c *Controller) parsePacketIn(pktIn *ofctrl.PacketIn) (*opsv1alpha1.Tracefl
+
+	if firstPacket {
+		// Uinstall the OVS flows after receiving the first packet.
+		c.ofClient.UninstallTraceflowFlows(tag)
```
Yes, that is true.
I am not sure it is theoretically possible that the flow deletion can be realized before all the packets from the same pipeline are sent, but even that is possible at least it is not quite possible in reality right?

antoninbas replied:
I think it is highly unlikely (if the packet was not recirculated in the datapath, I think it would be impossible, but I'm no OVS expert), so we can probably keep your current code but add a comment.

gran-vmv replied:
As implementation in this patch, I think the flow deletion will be always after packets processed in unicast environment, but this may not correct in multicast if we want to support it in future.
Meanwhile, sometimes Traceflow can capture unexpected network issues, such as duplicate packet, thus we may not need to delete the flows immediately. A hard timeout is enough.

jianjuns replied:
@gran-vmv : I changed to delete flows mainly for live-traffic case to avoid many packets of different connections being captured. But I got your point about packet dup, etc., so I changed the code to delete flows only for live-traffic Traceflow.

gran-vmv replied:
Sounds good.

jianjuns commented on pkg/agent/controller/traceflow/packetin.go:
```diff
@@ -89,17 +90,28 @@ func (c *Controller) parsePacketIn(pktIn *ofctrl.PacketIn) (*opsv1alpha1.Tracefl
+		c.ofClient.UninstallTraceflowFlows(tag)
+	}
+
+	tf, err := c.traceflowLister.Get(tfState.name)
```
You are right. I changed to return an error if not exists.

jianjuns commented on pkg/agent/openflow/client.go:
```diff
@@ -224,29 +224,13 @@ type Client interface {
-		icmpSequence uint16,
-		inPort uint32,
-		outPort int32) error
+	SendTraceflowPacket(dataplaneTag uint8, packet *binding.Packet, inPort uint32, outPort int32) error
+
+	// InstallTraceflowFlows installs flows for a Traceflow request.
+	InstallTraceflowFlows(dataplaneTag uint8, packet *binding.Packet, srcOFPort uint32, timeoutSeconds uint16) error
```
I agree. As discussed earlier, I think we should contract the TF packet in TF controller, and just call OF Client to send it out. But let us refactor later together with other packet out cases.

jianjuns commented on pkg/apis/ops/v1alpha1/types.go:
```diff
@@ -103,9 +103,15 @@ type Traceflow struct {
+	Source      Source       `json:"source,omitempty"`
+	Destination *Destination `json:"destination,omitempty"`
+	Packet      *Packet      `json:"packet,omitempty"`
+	// LiveTraffic indicates the Traceflow is to trace the live traffic
```
Yes, I plan to add a struct called CapturedPacket to TraceflowStatus.

jianjuns commented on pkg/apis/ops/v1alpha1/types.go:
```diff
@@ -103,9 +103,15 @@ type Traceflow struct {
+	// the first connection that matches the packet spec will be traced.
+	LiveTraffic bool `json:"liveTraffic,omitempty"`
+	// Timeout specifies the timeout of the Traceflow in seconds.
+	Timeout uint16 `json:"timeout,omitempty"`
```
Sure.

jianjuns commented on pkg/agent/controller/traceflow/traceflow_controller.go:
```diff
@@ -43,13 +43,11 @@ import (
 	"github.com/vmware-tanzu/antrea/pkg/querier"
 )
 
-type icmpType uint8
-type icmpCode uint8
```
We do not define a type for all other fields (IPFlags, TCPFlags, ICMPEchoSequence, IPProtocol, TPPort, etc.), I do not see why we define a type just for these three (and convert to uint8 later).

jianjuns commented on pkg/agent/controller/traceflow/traceflow_controller.go:
```diff
@@ -312,144 +364,176 @@ func (c *Controller) validateTraceflow(tf *opsv1alpha1.Traceflow) error {
+				if err != nil {
+					return nil, fmt.Errorf("failed to get the destination Pod: %v", err)
+				}
+				// DestinationMAC is nic here, will be set to Gateway
```
Yes. Fixed.

jianjuns commented:
I reverted the Traceflow API change of making Destination and Packet optional. Also added e2e tests for live-traffic Traceflow. But still need to fix inter-node e2e test failures.

jianjuns commented:
/test-e2e

jianjuns commented:
Debugging Kind e2e test failure..

tnqn commented on pkg/antctl/raw/traceflow/command.go:
```diff
@@ -369,7 +394,7 @@ func jsonOutput(r *Response) error {
 
 func getTFName(prefix string) string {
-	if !option.waiting {
+	if !option.nowait {
```
So this is actually a fix that a nowaiting traceflow request should have random suffix to avoid conflict?

jianjuns replied:
Actually, I am not sure about the right expectation here. @gran-vmv : could you explain why just no-waiting should have a prefix?

gran-vmv replied:
@lzhecheng Could you explain about this?

jianjuns replied:
I reverted my change anyway.

tnqn commented on pkg/apis/ops/v1alpha1/types.go:
```diff
@@ -185,7 +192,6 @@ type TCPHeader struct {
 
 // Packet includes header info.
 type Packet struct {
-	// TODO: change type IPHeader to *IPHeader and correct all internal references
```
Out of curiosity, don't we want to change it to pointer in the future?

jianjuns replied:
I think so, if you want I can add back the comment, but there are multiple fields I want to change and hope not to add too many TODOs.

tnqn replied:
No strong opinion, just to confirm if it's the case

tnqn commented on pkg/controller/traceflow/controller.go:
```diff
@@ -62,12 +62,13 @@ const (
 	traceflowTimeout = "Traceflow timeout"
+
+	// Traceflow timeout period.
+	defaultTimeoutDuration = 20 * time.Second
```
maybe we could move the default timeout to pkg/apis/ops/v1alpha1/ so antctl, controller, and agent can use same default value?

jianjuns replied:
Sounds a good idea, but do you want to add to types.go?

tnqn commented on pkg/ovs/openflow/ofctrl_action.go:
```diff
@@ -244,6 +244,11 @@ func (a *ofFlowAction) LoadPktMarkRange(value uint32, rng Range) FlowBuilder {
 	return a.LoadRange(NxmFieldPktMark, uint64(value), rng)
 }
 
+// LoadIPDSCP is an action to load data to IP DSP bits.
```
```suggestion
// LoadIPDSCP is an action to load data to IP DSCP bits.
```

jianjuns replied:
Thank!

jianjuns commented on pkg/agent/controller/traceflow/traceflow_controller.go:
```diff
@@ -270,27 +287,65 @@ func (c *Controller) startTraceflow(tf *opsv1alpha1.Traceflow) error {
+			matchPacket = packet
+		}
+		klog.V(2).Infof("Traceflow packet %v", *packet)
+		klog.Infof("XXX Traceflow packet %v", *packet)
```
Sorry, I am debugging an e2e test failure happens only with github Kind (cannot reproduce in my local env and local Kind), so added these logs. Please ignore all logs/code with "XXX".

jianjuns commented on pkg/antctl/raw/traceflow/command.go:
```diff
@@ -79,22 +83,39 @@ func init() {
 	Command.Flags().StringVarP(&option.outputType, "output", "o", "yaml", "output type: yaml (default), json")
-	Command.Flags().BoolVarP(&option.waiting, "wait", "", true, "if false, command returns without retrieving results")
 	Command.Flags().StringVarP(&option.flow, "flow", "f", "", "specify the flow (packet headers) of the Traceflow packet, including tcp_src, tcp_dst, tcp_flags, udp_src, udp_dst, ipv6")
+	Command.Flags().BoolVarP(&option.liveTraffic, "live-traffic", "", false, "if set, the Traceflow will trace the first packet of the matched live traffic flow")
```
Yes, thought about it, and feel no need to add a shorthand for users to remember. But I am open to add it too.

tnqn replied:
I think user will see the usage anyway when trying to start a live traffic traceflow, a shorthand will be convenient to type.

jianjuns replied:
Sure. Added -L.

jianjuns commented on pkg/agent/openflow/pipeline.go:
```diff
@@ -26,6 +26,8 @@ import (
 	"k8s.io/client-go/tools/cache"
 	"k8s.io/klog"
 
+	"github.com/contiv/libOpenflow/protocol"
```
Done.

jianjuns commented on pkg/agent/openflow/pipeline.go:
```diff
@@ -695,42 +697,96 @@ func (c *client) kubeProxyFlows(category cookie.Category) []binding.Flow {
+		// Match transport header
+		switch packet.IPProto {
+		case protocol.Type_ICMP, protocol.Type_IPv6ICMP:
+			flowBuilder = flowBuilder.MatchProtocol(binding.ProtocolICMPv6)
```
Yes, I found it in tests too. Corrected already.

jianjuns commented on pkg/agent/openflow/pipeline.go:
```diff
@@ -897,23 +954,28 @@ func (c *client) traceflowL2ForwardOutputFlows(dataplaneTag uint8, category cook
+			if liveTraffic {
+				// Clear the loaded DSCP bits before output.
+				flowBuilder = flowBuilder.Action().LoadIPDSCP(0).
+					Action().OutputRegRange(int(PortCacheReg), ofPortRegRange)
```
Sure.

jianjuns commented on pkg/antctl/raw/traceflow/command.go:
```diff
@@ -39,14 +39,18 @@ import (
 	clientset "github.com/vmware-tanzu/antrea/pkg/client/clientset/versioned"
 )
 
+const defaultTimeout time.Duration = time.Second * 10
```
Yes, but I like to timeout faster in antctl. A command hanging for 20 seconds seems too long to me.

jianjuns commented on pkg/antctl/raw/traceflow/command.go:
```diff
@@ -79,22 +83,39 @@ func init() {
-  Start a Traceflow from busybox0 to busybox1, with TCP header and 80 as destination port
+  Start a Traceflow from busybox0 to busybox1, with a TCP packet to destination port 80
   $antctl traceflow -S busybox0 -D busybox1 -f tcp,tcp_dst=80
+  Start a Traceflow for live TCP traffic from busybox0 to TCP port 90, with 1 minute timeout
```
Fixed.

jianjuns commented on pkg/agent/controller/traceflow/traceflow_controller.go:
```diff
@@ -184,7 +196,13 @@ func (c *Controller) updateTraceflow(_, curObj interface{}) {
 	tf := old.(*opsv1alpha1.Traceflow)
 	klog.Infof("Processing Traceflow %s DELETE event", tf.Name)
-	c.deallocateTag(tf)
+	tfState := c.deleteTraceflowState(tf)
```
We used to run into some race bugs, so I want not be a little careful here to change the code. Could you suggest the exact change you want - in syncTraceflow(), should I check or ignore c.traceflowLister.Get(traceflowName) returns before delete the cached state?

tnqn replied:
In my mind it could be:
```
tf, err := c.traceflowLister.Get(traceflowName)
if err != nil {
	if apierrors.IsNotFound(err) {
		tfState := c.deleteTraceflowState(traceflowName)
		if tfState != nil {
			err := c.ofClient.UninstallTraceflowFlows(tfState.tag)
			if err != nil {
				klog.Errorf("Failed to uninstall Traceflow %s flows: %v", tf.Name, err)
			}
		}
		return nil
	}
	return err
}
```

jianjuns replied:
So, even for delete, the object is still in client cache (can be got from traceflowLister)?

Changed. Check.

tnqn replied:
No, after it's deleted, Lister doesn't have it, that's why we should check if the error is "NotFound". I don't see we need the Traceflow object to remove the traceflow state cache. I think it only needs the name of the traceflow?

jianjuns replied:
Ok. Updated.

jianjuns commented:
/test-e2e

tnqn commented on pkg/agent/openflow/client.go:
```diff
@@ -15,13 +15,13 @@
 
 	"github.com/contiv/libOpenflow/protocol"
 	"github.com/contiv/ofnet/ofctrl"
+
```
klog should be in same group as ofnet

jianjuns replied:
Fixed.

tnqn commented on pkg/agent/openflow/pipeline.go:
```diff
@@ -23,6 +23,8 @@ import (
 	"time"
 
+	"github.com/contiv/libOpenflow/protocol"
+
```
ditto

jianjuns replied:
Fixed.

tnqn reviewed (approved):
LGTM, just two format issues.

tnqn reviewed (approved):
LGTM

jianjuns commented:
/test-all

jianjuns commented:
/test-all

jianjuns commented:
/test-all

jianjuns commented:
/test-networkpolicy
